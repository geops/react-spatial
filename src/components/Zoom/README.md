#

This demonstrates the use of Zoom.

```jsx
import React from 'react';
import Zoom from 'react-spatial/components/Zoom';
import BasicMap from 'react-spatial/components/BasicMap';
import OLMap from 'ol/Map';
import ConfigReader from 'react-spatial/ConfigReader';

const map = new OLMap({ controls: [] });
const layers = ConfigReader.readConfig(treeData);

function ZoomExample() {
  return (
    <div className="rs-zoom-example">
      <BasicMap map={map} layers={layers} tabIndex={0} />
      <Zoom map={map} zoomSlider />
    </div>
  );
};

<ZoomExample />;
```
