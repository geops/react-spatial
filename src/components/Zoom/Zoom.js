import React, {
  useEffect,
  useRef,
  useCallback,
  useState,
  useMemo,
} from 'react';
import PropTypes from 'prop-types';
import { FaPlus, FaMinus } from 'react-icons/fa';
import { ZoomSlider } from 'ol/control';
import OLMap from 'ol/Map';

const propTypes = {
  /**
   * An ol map.
   */
  map: PropTypes.instanceOf(OLMap).isRequired,

  /**
   * The zoom delta applied on each click.
   */
  delta: PropTypes.number,

  /**
   * Titles HTML attribtues for button.
   */
  titles: PropTypes.shape({
    zoomIn: PropTypes.string,
    zoomOut: PropTypes.stirng,
  }),

  /**
   * Children content of the zoom in button.
   */
  zoomInChildren: PropTypes.node,

  /**
   * Children content of the zoom out button.
   */
  zoomOutChildren: PropTypes.node,

  /**
   * Display a slider to zoom.
   */
  zoomSlider: PropTypes.bool,
};

const defaultProps = {
  titles: {
    zoomIn: 'Zoom in',
    zoomOut: 'Zoom out',
  },
  zoomInChildren: <FaPlus focusable={false} />,
  zoomOutChildren: <FaMinus focusable={false} />,
  zoomSlider: false,
  delta: 1,
};

const updateZoom = (map, zoomAction) => {
  map.getView().cancelAnimations();
  const zoom = map.getView().getZoom();

  map.getView().animate({ zoom: zoom + zoomAction });
};

/**
 * This component creates a zoom wrapper.
 */
function Zoom({
  map,
  titles,
  zoomInChildren,
  zoomOutChildren,
  zoomSlider,
  delta,
  ...other
}) {
  const ref = useRef();
  const [currentZoom, setZoom] = useState();

  const zoomIn = useCallback(
    (evt) => {
      if (!evt.which || evt.which === 13) {
        updateZoom(map, delta);
      }
    },
    [map],
  );

  const zoomOut = useCallback(
    (evt) => {
      if (!evt.which || evt.which === 13) {
        updateZoom(map, -delta);
      }
    },
    [map],
  );

  const zoomInDisabled = useMemo(() => {
    return (
      currentZoom >=
      map.getView().getConstrainedZoom(map.getView().getMaxZoom())
    );
  }, [currentZoom]);

  const zoomOutDisabled = useMemo(() => {
    return (
      currentZoom <=
      map.getView().getConstrainedZoom(map.getView().getMinZoom())
    );
  }, [currentZoom]);

  useEffect(() => {
    /* Trigger zoom update to disable zooms on max and min */
    const zoomListener = () => setZoom(map.getView().getZoom());
    map.on('moveend', zoomListener);

    let control;
    if (zoomSlider && ref.current) {
      control = new ZoomSlider();
      // We don't want to navigate to the zoom slider using TAB navigation.
      control.element.firstElementChild.tabIndex = -1;
      // Set the zoom slider in the custom control wrapper.
      control.setTarget(ref.current);
      map.addControl(control);
    }
    return () => {
      map.un('moveend', zoomListener);
      if (control) {
        map.removeControl(control);
      }
    };
  }, [map, zoomSlider, ref.current]);

  return (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <div className="rs-zooms-bar" {...other}>
      <button
        type="button"
        tabIndex={0}
        className="rs-zoom-in"
        title={titles.zoomIn}
        onClick={zoomIn}
        onKeyPress={zoomIn}
        disabled={zoomInDisabled}
      >
        {zoomInChildren}
      </button>
      {zoomSlider ? <div className="rs-zoomslider-wrapper" ref={ref} /> : null}
      <button
        type="button"
        tabIndex={0}
        className="rs-zoom-out"
        title={titles.zoomOut}
        onClick={zoomOut}
        onKeyPress={zoomOut}
        disabled={zoomOutDisabled}
      >
        {zoomOutChildren}
      </button>
    </div>
  );
}

Zoom.propTypes = propTypes;
Zoom.defaultProps = defaultProps;

export default React.memo(Zoom);
