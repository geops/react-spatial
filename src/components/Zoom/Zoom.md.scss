.rs-zoom-example {
  position: relative;

  .rs-map {
    height: 400px;
  }

  .rs-zooms-bar {
    top: 0;
    right: 10px;
  }

  .rs-zoom-in,
  .rs-zoom-out {
    background: $btn-secondary-color;
    border-radius: 50%;
    color: white;
  }
}
