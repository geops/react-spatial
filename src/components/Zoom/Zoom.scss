.rs-zooms-bar {
  position: absolute;
  top: 0;
  right: 0;

  div {
    margin: 10px 0;
  }

  .rs-zoom-in,
  .rs-zoom-out {
    cursor: pointer;
    height: $btn-size-base;
    width: $btn-size-base;
    display: flex;
    justify-content: center;
    align-items: center;
    transition: background-color 0.5s ease, color 0.5s ease;
    border: none;

    &:disabled {
      pointer-events: none;
      opacity: 0.4;
    }
  }

  .rs-zoomslider-wrapper {
    display: flex;
    justify-content: center;

    .ol-zoomslider {
      position: relative;
      top: 0;
      left: 0;
      user-select: none;
      background-color: rgba(255, 255, 255, 0.4);
      border: 1px solid $brand-secondary;
      border-radius: 3px;
      outline: none;
      overflow: hidden;
      width: 14px;
      height: 200px;
      margin: 0;

      .ol-zoomslider-thumb {
        user-select: none;
        position: relative;
        display: block;
        background: $brand-secondary;
        outline: none;
        overflow: hidden;
        cursor: pointer;
        font-size: 1.14em;
        height: 20px;
        width: 14px;
        margin: 0;
        padding: 0;
        border: none;
      }
    }
  }
}
