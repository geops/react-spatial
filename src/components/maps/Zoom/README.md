#

This demonstrates the use of Zoom.

```jsx
import React from 'react';
import Zoom from 'react-spatial/components/maps/Zoom';
import BasicMap from 'react-spatial/components/maps/BasicMap';
import OLMap from 'ol/Map';
import ConfigReader from 'react-spatial/ConfigReader';

const map = new OLMap({ controls: [] });
const layers = ConfigReader.readConfig(treeData);

function ZoomExample() {
  return (
    <div className="rs-zoom-example">
      <BasicMap map={map} layers={layers} tabIndex={0} />
      <Zoom map={map} zoomSlider />
    </div>
  );
};

<ZoomExample />;
```
