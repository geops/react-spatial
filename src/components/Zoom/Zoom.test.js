import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { act } from 'react-dom/test-utils';
import MapEvent from 'ol/MapEvent';
import OLView from 'ol/View';
import OLMap from 'ol/Map';
import Zoom from './Zoom';

configure({ adapter: new Adapter() });

describe('Zoom', () => {
  test('should match snapshot.', () => {
    const map = new OLMap({});
    const component = renderer.create(<Zoom map={map} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should match snapshot with custom attributes', () => {
    const map = new OLMap({});
    const component = renderer.create(
      <Zoom map={map} className="foo" tabIndex={-1} title="bar" />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should match snapshot with zoom slider', () => {
    const map = new OLMap({});
    const component = renderer.create(<Zoom map={map} zoomSlider />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  [
    ['click', {}],
    ['keypress', { which: 13 }],
  ].forEach((evt) => {
    test(`should zoom in on ${evt[0]}.`, () => {
      const map = new OLMap({ view: new OLView({ zoom: 5 }) });
      const zooms = shallow(<Zoom map={map} />);
      zooms
        .find('.rs-zoom-in')
        .first()
        .simulate(...evt);

      expect(map.getView().getZoom()).toBe(6);
    });

    test(`should zoom in on ${evt[0]} (delta: 0.3).`, () => {
      const map = new OLMap({ view: new OLView({ zoom: 5 }) });
      const zooms = shallow(<Zoom map={map} delta={0.3} />);
      zooms
        .find('.rs-zoom-in')
        .first()
        .simulate(...evt);

      expect(map.getView().getZoom()).toBe(5.3);
    });

    test(`should zoom out on ${evt[0]}.`, () => {
      const map = new OLMap({ view: new OLView({ zoom: 5 }) });
      const zooms = shallow(<Zoom map={map} />);
      zooms
        .find('.rs-zoom-out')
        .first()
        .simulate(...evt);

      expect(map.getView().getZoom()).toBe(4);
    });

    test(`should zoom out on ${evt[0]} (delta: 0.3).`, () => {
      const map = new OLMap({ view: new OLView({ zoom: 5 }) });
      const zooms = shallow(<Zoom map={map} delta={0.3} />);
      zooms
        .find('.rs-zoom-out')
        .first()
        .simulate(...evt);

      expect(map.getView().getZoom()).toBe(4.7);
    });
  });

  test('should remove zoomSlider control on unmount.', () => {
    const map = new OLMap({});
    const spy = jest.spyOn(map, 'removeControl');
    const spy2 = jest.spyOn(map, 'addControl');
    const wrapper = mount(<Zoom map={map} zoomSlider />);
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.unmount();
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][0]).toBe(spy2.mock.calls[0][0]);
  });

  test('should disable zoom-in button on mount with max zoom..', () => {
    const map = new OLMap({
      view: new OLView({ maxZoom: 20, zoom: 20 }),
    });
    const spy = jest.spyOn(map.getView(), 'setZoom');
    const wrapper = mount(<Zoom map={map} />);
    act(() => {
      map.dispatchEvent(new MapEvent('moveend', map));
    });
    wrapper.update();
    expect(wrapper.find('.rs-zoom-in').prop('disabled')).toEqual(true);
    wrapper.find('.rs-zoom-in').first().simulate('click');
    expect(spy).toHaveBeenCalledTimes(0);
  });

  test('should disable zoom-out button on mount with min zoom.', () => {
    const map = new OLMap({
      view: new OLView({ minZoom: 2, zoom: 2 }),
    });
    const spy = jest.spyOn(map.getView(), 'setZoom');
    const wrapper = mount(<Zoom map={map} />);
    act(() => {
      map.dispatchEvent(new MapEvent('moveend', map));
    });
    wrapper.update();
    expect(wrapper.find('.rs-zoom-out').prop('disabled')).toEqual(true);
    wrapper.find('.rs-zoom-out').first().simulate('click');
    expect(spy).toHaveBeenCalledTimes(0);
  });
});

test('should disable zoom-out button when reaching min zoom.', () => {
  const map = new OLMap({
    view: new OLView({ minZoom: 2, zoom: 3 }),
  });
  const spy = jest.spyOn(map.getView(), 'setZoom');
  const wrapper = mount(<Zoom map={map} />);
  wrapper.find('.rs-zoom-out').first().simulate('click');
  expect(spy).toHaveBeenCalledTimes(1);
  act(() => {
    map.dispatchEvent(new MapEvent('moveend', map));
  });
  wrapper.update();
  expect(wrapper.find('.rs-zoom-out').prop('disabled')).toEqual(true);
});
