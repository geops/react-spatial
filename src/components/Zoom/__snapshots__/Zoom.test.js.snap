// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zoom should match snapshot with custom attributes 1`] = `
<div
  className="foo"
  tabIndex={-1}
  title="bar"
>
  <button
    className="rs-zoom-in"
    disabled={false}
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
    title="Zoom in"
    type="button"
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      />
    </svg>
  </button>
  <button
    className="rs-zoom-out"
    disabled={false}
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
    title="Zoom out"
    type="button"
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      />
    </svg>
  </button>
</div>
`;

exports[`Zoom should match snapshot with zoom slider 1`] = `
<div
  className="rs-zooms-bar"
>
  <button
    className="rs-zoom-in"
    disabled={false}
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
    title="Zoom in"
    type="button"
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      />
    </svg>
  </button>
  <div
    className="rs-zoomslider-wrapper"
  />
  <button
    className="rs-zoom-out"
    disabled={false}
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
    title="Zoom out"
    type="button"
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      />
    </svg>
  </button>
</div>
`;

exports[`Zoom should match snapshot. 1`] = `
<div
  className="rs-zooms-bar"
>
  <button
    className="rs-zoom-in"
    disabled={false}
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
    title="Zoom in"
    type="button"
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      />
    </svg>
  </button>
  <button
    className="rs-zoom-out"
    disabled={false}
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
    title="Zoom out"
    type="button"
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      />
    </svg>
  </button>
</div>
`;
